components/places/ios/Places/Places.swift

```diff
@@ -0,0 +1,418 @@
+        return try queue.sync {
+            try self.checkApi()
+            let buffer = try PlacesError.unwrap { error in
+                bookmarks_get_tree(self.handle, rootGUID, error)
+            }
+            if buffer.data == nil {
+                return nil
+            }
+            defer { places_destroy_bytebuffer(buffer) }
+            // This should never fail, since we encoded it on the other side with Rust,
+            // should we use `try! instead?
+            let msg = try MsgTypes_BookmarkNode(serializedData: Data(placesRustBuffer: buffer))
+            return unpackProtobuf(msg: msg, expectChildren: true)
+        }
+    }
+
+    /**
+     * Returns the information about the bookmark with the provided id.
+     *
+     * This differs from `getBookmarksTree` in that it does not populate the `children` list
+     * if `guid` refers to a folder (However, it's `childGUIDs` list will be
+     * populated).
+     *
+     * - Parameter guid: the guid of the bookmark to fetch.
+     *
+     * - Returns: The bookmark node, or null if the provided guid didn't refer to a
+     *            known bookmark item.
+     */
+    func getBookmark(guid: String) throws -> BookmarkNode? {
+        return try queue.sync {
+            try self.checkApi()
+            let buffer = try PlacesError.unwrap { error in
+                bookmarks_get_by_guid(self.handle, guid, error)
+            }
+            if buffer.data == nil {
+                return nil
+            }
+            defer { places_destroy_bytebuffer(buffer) }
+            // Should this be `try!`?
+            let msg = try MsgTypes_BookmarkNode(serializedData: Data(placesRustBuffer: buffer))
+            return unpackProtobuf(msg: msg, expectChildren: false)
+        }
+    }
+
+    /**
+     * Returns the list of bookmarks with the provided URL.
+     *
+     * - Note: If the URL is not percent-encoded/punycoded, that will be performed
+     *         internally, and so the returned bookmarks may not have an identical
+     *         URL to the one passed in, however, it will be the same according to
+     *         https://url.spec.whatwg.org
+     *
+     * - Parameter url: The url to search for.
+     *
+     * - Returns: A list of bookmarks that have the requested URL.
+     */
+    func getBookmarksWithURL(url: String) throws -> [BookmarkNode] {
+        return try queue.sync {
+            try self.checkApi()
+            let buffer = try PlacesError.unwrap { error in
+                bookmarks_get_all_with_url(self.handle, url, error)
+            }
+            defer { places_destroy_bytebuffer(buffer) }
+            // Should this be `try!`?
+            let msg = try MsgTypes_BookmarkNodeList(serializedData: Data(placesRustBuffer: buffer))
+            return unpackProtobufList(msg: msg)
+        }
+    }
+
+}
+
+/**
+ * A read-write connection to the places database.
+ */
+public class PlacesWriteConn : PlacesReadConn {
```

Comment:
nit: `PlacesWriteConnection`

Reply:
I really like how you've separated out the read-only methods from the writable functions 👍